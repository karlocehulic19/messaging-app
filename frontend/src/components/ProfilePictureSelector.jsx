import { useDropzone } from "react-dropzone";
import styles from "./styles/ProfilePictureSelector.module.css";
import usePictureURL from "../hooks/usePictureURL";

function ProfilePictureSelector() {
  const { acceptedFiles, fileRejections, getRootProps, getInputProps } =
    useDropzone({
      maxFiles: 1,
      accept: {
        "image/jpeg": [],
        "image/png": [],
      },
    });

  const pictureURL = usePictureURL(acceptedFiles[0]);

  const fileRejected = !!fileRejections.length;

  return (
    <div
      data-testid="profile-picture-container"
      className="profile-picture-container"
      id={
        pictureURL
          ? styles["picture-container-demo"]
          : styles["picture-container"]
      }
    >
      {pictureURL && (
        <div aria-label="Profile picture demonstration">
          <label htmlFor="profile-picture">Profile picture</label>
          <img src={pictureURL} alt="Current Profile picture" />
        </div>
      )}

      <div className="picture-dropbox-container">
        {fileRejected && (
          <p>Please select image that is in PNG or JPEG format</p>
        )}
        <div
          {...getRootProps({
            className: "dropbox",
            id: styles["picture-dropbox"],
            ["aria-label"]: "Profile picture dropbox",
          })}
        >
          <input
            {...getInputProps({
              htmlFor: "profile-picture",
              id: "profile-picture",
              ["data-testid"]: "picture-input",
            })}
          />
          <p>Drop your profile picture here!</p>
        </div>
      </div>
    </div>
  );
}

export default ProfilePictureSelector;
