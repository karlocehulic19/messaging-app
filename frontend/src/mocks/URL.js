global.URL.createObjectURL = vi.fn(() => "mockedFilePath");
global.URL.revokeObjectURL = vi.fn(() => null);
