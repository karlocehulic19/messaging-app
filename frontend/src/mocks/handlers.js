import { http, HttpResponse } from "msw";
import { config } from "../Constants";

export const handlers = [
  http.post(`${config.url.BACKEND_URL}/login`, async ({ request }) => {
    const body = await request.json();
    if (!body.username || !body.password) {
      return new HttpResponse("Missing credentials", { status: 401 });
    }

    if (body.username != "someUsername" || body.password != "somePassword") {
      return HttpResponse.json(
        { messages: ["Username or password is incorrect"] },
        { status: 401 }
      );
    }
    return HttpResponse.json({ token: "randomJWTtoken" }, { status: 200 });
  }),
];
