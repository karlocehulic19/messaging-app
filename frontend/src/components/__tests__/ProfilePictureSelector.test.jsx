import { cleanup, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ProfilePictureSelector from "../ProfilePictureSelector";
import "../../mocks/URL";
import { expect } from "vitest";

const mockedPNGFile = new File(["test"], "test.png", { type: "image/png" });
const mockedJPEGFile = new File(["test"], "test.jpeg", { type: "image/jpeg" });
const mockedInvalidFile = new File(["test"], "test.txt", {
  type: "application/json",
});

const setup = (userEventOptions = { applyAccept: false }) => {
  const mockedFormatter = vi.fn(() => "mockBase64");
  render(<ProfilePictureSelector imageFormatter={mockedFormatter} />);

  return { user: userEvent.setup(userEventOptions), mockedFormatter };
};

const setupInvalidFile = async () => {
  const { user } = setup();

  await user.upload(screen.getByTestId("picture-input"), mockedInvalidFile);

  return { user };
};

const setupValidFile = async () => {
  const { user } = setup();

  await user.upload(screen.getByTestId("picture-input"), mockedPNGFile);

  return { user };
};

describe("<ProfilePictureSelector />", () => {
  it("displays error message on invalid file type", async () => {
    const { user } = setup();

    await user.upload(screen.getByTestId("picture-input"), mockedInvalidFile);

    expect(
      screen.getByText("Please select image that is in PNG or JPEG format")
    ).toBeInTheDocument();
  });

  it("doesn't displays error message on PNG file type", async () => {
    const { user } = setup();

    await user.upload(screen.getByTestId("picture-input"), mockedPNGFile);

    expect(
      screen.queryByText("Please select image that is in PNG or JPEG format")
    ).toBeNull();
  });

  it("doesn't displays error message on JPEG file type", async () => {
    const { user } = setup();

    await user.upload(screen.getByTestId("picture-input"), mockedJPEGFile);

    expect(
      screen.queryByText("Please select image that is in PNG or JPEG format")
    ).toBeNull();
  });

  it("displays profile picture with right file type", async () => {
    await setupValidFile();

    expect(
      screen.getByLabelText("Profile picture demonstration")
    ).toMatchSnapshot();
  });

  it("doesn't display profile picture if non are selected", () => {
    setup();

    expect(screen.queryByRole("img")).not.toBeInTheDocument();
  });

  it("uses demonstration styling when image is imported", async () => {
    await setupValidFile();

    expect(screen.getByTestId("profile-picture-container")).toMatchSnapshot();
  });

  it("doesn't use demonstration styling when image isn't imported", async () => {
    await setupInvalidFile();

    expect(screen.getByTestId("profile-picture-container")).toMatchSnapshot();

    cleanup();

    setup();

    expect(screen.getByTestId("profile-picture-container")).toMatchSnapshot();
  });

  it("doesn't remove profile image after invalid picture selection", async () => {
    const { user } = await setupValidFile();

    await user.upload(screen.getByTestId("picture-input"), mockedInvalidFile);
    expect(screen.getByRole("img")).toBeInTheDocument();
  });

  it("removes profile image after clicking on remove button", async () => {
    const { user } = await setupValidFile();

    await user.click(screen.getByText("Remove Picture"));

    expect(screen.queryByRole("img")).not.toBeInTheDocument();
  });
});
