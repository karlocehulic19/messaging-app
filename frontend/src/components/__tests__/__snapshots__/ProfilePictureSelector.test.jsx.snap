// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ProfilePictureSelector /> > displays profile picture with right file type 1`] = `
<div
  aria-label="Profile picture demonstration"
>
  <label
    for="profile-picture"
  >
    Profile picture
  </label>
  <img
    alt="Current Profile picture"
    src="mockedFilePath"
  />
</div>
`;

exports[`<ProfilePictureSelector /> > doesn't use demonstration styling when image isn't imported 1`] = `
<div
  class="profile-picture-container"
  data-testid="profile-picture-container"
  id="_picture-container_ee6cb2"
>
  <div
    class="picture-dropbox-container"
  >
    <div
      aria-label="Profile picture dropbox"
      class="dropbox"
      id="_picture-dropbox_ee6cb2"
      role="presentation"
      tabindex="0"
    >
      <input
        accept="image/jpeg,image/png"
        data-testid="picture-input"
        for="profile-picture"
        id="profile-picture"
        multiple=""
        style="border: 0px; clip: rect(0px, 0px, 0px, 0px); clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
        tabindex="-1"
        type="file"
      />
      <p>
        Drop your profile picture here!
      </p>
    </div>
  </div>
</div>
`;

exports[`<ProfilePictureSelector /> > doesn't use demonstration styling when image isn't imported 2`] = `
<div
  class="profile-picture-container"
  data-testid="profile-picture-container"
  id="_picture-container_ee6cb2"
>
  <div
    class="picture-dropbox-container"
  >
    <div
      aria-label="Profile picture dropbox"
      class="dropbox"
      id="_picture-dropbox_ee6cb2"
      role="presentation"
      tabindex="0"
    >
      <input
        accept="image/jpeg,image/png"
        data-testid="picture-input"
        for="profile-picture"
        id="profile-picture"
        multiple=""
        style="border: 0px; clip: rect(0px, 0px, 0px, 0px); clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
        tabindex="-1"
        type="file"
      />
      <p>
        Drop your profile picture here!
      </p>
    </div>
  </div>
</div>
`;

exports[`<ProfilePictureSelector /> > uses demonstration styling when image is imported 1`] = `
<div
  class="profile-picture-container"
  data-testid="profile-picture-container"
  id="_picture-container-demo_ee6cb2"
>
  <div
    aria-label="Profile picture demonstration"
  >
    <label
      for="profile-picture"
    >
      Profile picture
    </label>
    <img
      alt="Current Profile picture"
      src="mockedFilePath"
    />
  </div>
  <div
    class="picture-dropbox-container"
  >
    <div
      aria-label="Profile picture dropbox"
      class="dropbox"
      id="_picture-dropbox_ee6cb2"
      role="presentation"
      tabindex="0"
    >
      <input
        accept="image/jpeg,image/png"
        data-testid="picture-input"
        for="profile-picture"
        id="profile-picture"
        multiple=""
        style="border: 0px; clip: rect(0px, 0px, 0px, 0px); clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
        tabindex="-1"
        type="file"
      />
      <p>
        Drop your profile picture here!
      </p>
    </div>
  </div>
</div>
`;
